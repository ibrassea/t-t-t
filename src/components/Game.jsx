import React, { Component } from "react";
import { observable, action } from "mobx";
import { observer } from "mobx-react";
import { Button, Col, Row, Container, Card,
  CardTitle, CardSubtitle, CardText, CardImg,
  InputGroup, InputGroupAddon, InputGroupText, Input,
  CardBody} from 'reactstrap';
import Pieces from "./Pieces";
import WinLine from "./WinLine";



@observer
class Game extends React.Component {
  constructor(){
    super();
    this.barCollapsed = false;
  }

//returns the pieces on the board depending on turn and mode.
getCell(cellIndex){
  let pieceValue = this.props.store.board[cellIndex]
  return Pieces(pieceValue, this.props.store.gameMode, this.props.store.winningLine, cellIndex, this.props.store.winningLineIndex);

 }

 gameOver(){
   return this.props.store.isGameOver;
 }

toggleModeClick(){
  this.props.store.toggleMode();
}
resetGameClick(){
  this.props.store.resetGame();
}

boardClick(e){
  console.log("board clicked at", e.target.id )
  store.makeMove(e.target.id)
}

toggleNavbar(){
    this.barCollapsed = !this.barCollapsed;
}


  render() {
    return (
    <Container>


        <Row>
          <Col lg="8">
            <Card className="mt-4" >
            <Row>
              <Col lg="12" className=" m-3">
              Next turn:   <span className={(this.props.store.playerOneTurn)? 'player active':'player'} > Player 1: O </span>
            <span className={ (!this.props.store.playerOneTurn)? 'player active':'player'}> Player 2: X</span>
              </Col>
            </Row>
          </Card>

          <Card className="mt-4">
          <div className="wrapper">

            <div className="board" id="board" onClick={this.boardClick}>
              <div id="0" className="tic-cell blt btt">{this.getCell(0)}</div>
              <div id="1" className="tic-cell btt">{this.getCell(1)}
              </div>
              <div id="2" className="tic-cell btt brt">{this.getCell(2)}</div>
              <div id="3" className="tic-cell blt">{this.getCell(3)}
              </div>
              <div id="4" className="tic-cell">{this.getCell(4)}
              </div>
              <div id="5" className="tic-cell brt">{this.getCell(5)}</div>
              <div id="6" className="tic-cell blt bbt">{this.getCell(6)}
              </div>
              <div id="7" className="tic-cell bbt">{this.getCell(7)}</div>
              <div id="8" className="tic-cell bbt brt">{this.getCell(8)}</div>

          </div>
      </div>
      </Card>
  </Col>

  <Col md="4">

  <Card className="mt-4" >
        <CardBody >
          <CardTitle>Game options:  {this.props.store.isGameOver && <strong> {this.props.store.winner}</strong>}</CardTitle>



      <InputGroup>
        <InputGroupAddon addonType="prepend">
          <InputGroupText>
            <Input addon type="checkbox" aria-label="Checkbox for following text input" onClick={()=>{this.toggleModeClick()}}/>
          </InputGroupText>
        </InputGroupAddon>
        <Input placeholder="Zardoz Mode" disabled />


      </InputGroup>

     <Row className="pt-4">
      <Col xs="4">
         <Button color="primary" onClick={()=>{this.resetGameClick()}}> New game </Button>
      </Col>
        <Col xs="4">
        <Button color="secondary" disabled={!this.props.store.moveHistory.length} onClick={()=>{this.props.store.undoMove()}}> Undo Last</Button>
        </Col>
     </Row>


        </CardBody>
      </Card>

  </Col>


  </Row>
  </Container>

  )};

}//class

export default Game;
