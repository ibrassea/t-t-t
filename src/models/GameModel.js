import { observable, computed, action } from "mobx";


export default class GameModel {

  constructor(){
    this.allPossibleLines = [[0,1,2], [3,4,5], [6,7,8],
                         [0,3,6], [1,4,7], [2,5,8],
                         [0,4,8], [2,4,6]];

                         /* index representation
                         0 = horizontal top
                         1 = horizontal middle
                         2 = horizontal bottom
                         3 = vertical left
                         4 = vertical middle
                         5 = vertical right
                         6 = diagonal left top to borrom right
                         7 = diagonal left bottom to right top
                        */


    this.playerOneTurn = true;
    this.pieceNames = {0: "Circles", 1: "Crosses"};
    this.winner = "";
  }

  @observable board = Array(9).fill(null);
  @observable gameMode = "DEFAULT" //or ZARDOZ
  @observable gameOver = false;
  @observable winningLine = [];
  @observable winningLineIndex = null;
  @observable moveHistory = []; // arr of move obj.

  @computed
  get isGameOver() {
    let gameOver = this.gameOver == true || !this.board.some( x => x == null )
    if(gameOver && this.winner == ""){
      this.winner = "Draw"

    }

    return gameOver
  }

  isLine(num){
    let isLine = false;
    this.allPossibleLines.forEach((lineIndices, index)=>{
      if (lineIndices.every( index => this.board[index] === num )){
        isLine = true;
        this.winningLine = lineIndices;
        this.winningLineIndex = index;

        return true;
      }
   });

   return isLine;
  }

  @action
  toggleMode(){
    this.gameMode =   this.gameMode == "DEFAULT" ? "ZARDOZ" : "DEFAULT";
  }

  @action
  resetGame(){
    this.board = Array(9).fill(null);
    this.gameOver = false;
    this.winner = "";
    this.winningLine = [];
    this.winningLineIndex = null;
    this.moveHistory = [];
  }

  @action
  undoMove(){

    if(this.moveHistory.length){
      this.gameOver = false;
      this.winner = "";
      this.winningLine = [];
      this.winningLineIndex = null;

      let lastMove = this.moveHistory.pop();
      this.board[lastMove.boardIndex] = null;
      this.playerOneTurn = !this.playerOneTurn;
    }
    //remove from history.


  }

  @action
  makeMove(boardIndex) {

    if(this.isGameOver){console.log("game over"); return}

    let isWinningMove = false;

    if(this.board[boardIndex] === null && boardIndex){
      this.board[boardIndex] = this.playerOneTurn ? 0: 1;

      this.moveHistory.push({
        boardIndex: boardIndex,
        playerTurn: this.board[boardIndex]
      })

      isWinningMove = this.isLine(this.board[boardIndex]);

      if(isWinningMove){
        this.gameOver = true;
        this.winner = this.pieceNames[this.board[boardIndex]] + " win!";
      }
      this.playerOneTurn = !this.playerOneTurn;
    }
  }
}
