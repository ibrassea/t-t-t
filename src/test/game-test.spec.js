
import React from 'react';
import renderer from 'react-test-renderer';
import Game from '../components/Game';
import GameModel from "../models/GameModel";

let store = new GameModel();
let component = renderer.create(
  <Game store={store} />
);

function resetGame(){
  store.resetGame();
}

function makeMoves(movesArray){
 movesArray.forEach((index)=>{
   store.makeMove(index);
 })
}


describe("Tic tac toe functional test", function() {

  afterEach(()=>{
    resetGame();
  });

  test('2 Snapshot test new game and with two moves', function() {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    //make some moves to change component for next snapshot.
    store.makeMove(1)
    store.makeMove(3)

    // re-rendering for snapshot.
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

  });

  test('Snapshot test with win', function() {
    makeMoves([1,3,4,6,7]);
    // re-rendering for snapshot.
   let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

  });

  test("Play a full game", ()=>{

    //checks game is in progress
    expect(store.isGameOver).toEqual(false);

    //make moves to win the game.
    makeMoves([1,3,4,6,7]);
    //checks game is done
    expect(store.isGameOver).toEqual(true);

  })

  test("Undo last move after win", ()=>{
    //make moves to win the game.
    makeMoves([1,3,4,6,7]);
    //checks game is done
    expect(store.isGameOver).toEqual(true);

    store.undoMove();
    expect(store.isGameOver).toEqual(false);

  })

  test("Reset game", ()=>{
    //make moves to win the game.
    makeMoves([1,3,4,6,7]);
    //checks game is done
    expect(store.isGameOver).toEqual(true);

    store.undoMove();
    expect(store.isGameOver).toEqual(false);

  })

  test("Zardoz game mode", ()=>{
    //make some moves
    makeMoves([1,3,4]);
    //checks game is done
    expect(store.gameMode).toEqual('DEFAULT');

    store.toggleMode();
    expect(store.gameMode).toEqual('ZARDOZ');

    store.undoMove();
    expect(store.isGameOver).toEqual(false);

  })


});
