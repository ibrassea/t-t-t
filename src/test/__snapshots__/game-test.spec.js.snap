// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tic tac toe functional test 2 Snapshot test new game and with two moves 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-lg-8"
    >
      <div
        className="mt-4 card"
      >
        <div
          className="row"
        >
          <div
            className=" m-3 col-lg-12"
          >
            Next turn:   
            <span
              className="player active"
            >
               Player 1: O 
            </span>
            <span
              className="player"
            >
               Player 2: X
            </span>
          </div>
        </div>
      </div>
      <div
        className="mt-4 card"
      >
        <div
          className="wrapper"
        >
          <div
            className="board"
            id="board"
            onClick={[Function]}
          >
            <div
              className="tic-cell blt btt"
              id="0"
            >
              
            </div>
            <div
              className="tic-cell btt"
              id="1"
            >
              
            </div>
            <div
              className="tic-cell btt brt"
              id="2"
            >
              
            </div>
            <div
              className="tic-cell blt"
              id="3"
            >
              
            </div>
            <div
              className="tic-cell"
              id="4"
            >
              
            </div>
            <div
              className="tic-cell brt"
              id="5"
            >
              
            </div>
            <div
              className="tic-cell blt bbt"
              id="6"
            >
              
            </div>
            <div
              className="tic-cell bbt"
              id="7"
            >
              
            </div>
            <div
              className="tic-cell bbt brt"
              id="8"
            >
              
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-md-4"
    >
      <div
        className="mt-4 card"
      >
        <div
          className="card-body"
        >
          <div
            className="card-title"
          >
            Game options:  
          </div>
          <div
            className="input-group"
          >
            <div
              className="input-group-prepend"
            >
              <span
                className="input-group-text"
              >
                <input
                  aria-label="Checkbox for following text input"
                  className=""
                  onClick={[Function]}
                  type="checkbox"
                />
              </span>
            </div>
            <input
              className="form-control"
              disabled={true}
              placeholder="Zardoz Mode"
              type="text"
            />
          </div>
          <div
            className="pt-4 row"
          >
            <div
              className="col-4"
            >
              <button
                aria-label={null}
                className="btn btn-primary"
                onClick={[Function]}
                type="button"
              >
                 New game 
              </button>
            </div>
            <div
              className="col-4"
            >
              <button
                aria-label={null}
                className="btn btn-secondary disabled"
                disabled={true}
                onClick={[Function]}
                type="button"
              >
                 Undo Last
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tic tac toe functional test 2 Snapshot test new game and with two moves 2`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-lg-8"
    >
      <div
        className="mt-4 card"
      >
        <div
          className="row"
        >
          <div
            className=" m-3 col-lg-12"
          >
            Next turn:   
            <span
              className="player active"
            >
               Player 1: O 
            </span>
            <span
              className="player"
            >
               Player 2: X
            </span>
          </div>
        </div>
      </div>
      <div
        className="mt-4 card"
      >
        <div
          className="wrapper"
        >
          <div
            className="board"
            id="board"
            onClick={[Function]}
          >
            <div
              className="tic-cell blt btt"
              id="0"
            >
              
            </div>
            <div
              className="tic-cell btt"
              id="1"
            >
              <svg
                className="circle piece"
              >
                <circle
                  cx="50%"
                  cy="50%"
                  fill="none"
                  r="45%"
                  stroke="#5a5a5a"
                  strokeWidth="3"
                />
              </svg>
            </div>
            <div
              className="tic-cell btt brt"
              id="2"
            >
              
            </div>
            <div
              className="tic-cell blt"
              id="3"
            >
              <svg
                className="cross piece"
              >
                <line
                  stroke="#f47521"
                  strokeWidth="5"
                  x1="0%"
                  x2="100%"
                  y1="0"
                  y2="100%"
                />
                <line
                  stroke="#f47521"
                  strokeWidth="5"
                  x1="0"
                  x2="100%"
                  y1="100%"
                  y2="0"
                />
              </svg>
            </div>
            <div
              className="tic-cell"
              id="4"
            >
              
            </div>
            <div
              className="tic-cell brt"
              id="5"
            >
              
            </div>
            <div
              className="tic-cell blt bbt"
              id="6"
            >
              
            </div>
            <div
              className="tic-cell bbt"
              id="7"
            >
              
            </div>
            <div
              className="tic-cell bbt brt"
              id="8"
            >
              
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-md-4"
    >
      <div
        className="mt-4 card"
      >
        <div
          className="card-body"
        >
          <div
            className="card-title"
          >
            Game options:  
          </div>
          <div
            className="input-group"
          >
            <div
              className="input-group-prepend"
            >
              <span
                className="input-group-text"
              >
                <input
                  aria-label="Checkbox for following text input"
                  className=""
                  onClick={[Function]}
                  type="checkbox"
                />
              </span>
            </div>
            <input
              className="form-control"
              disabled={true}
              placeholder="Zardoz Mode"
              type="text"
            />
          </div>
          <div
            className="pt-4 row"
          >
            <div
              className="col-4"
            >
              <button
                aria-label={null}
                className="btn btn-primary"
                onClick={[Function]}
                type="button"
              >
                 New game 
              </button>
            </div>
            <div
              className="col-4"
            >
              <button
                aria-label={null}
                className="btn btn-secondary"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                 Undo Last
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tic tac toe functional test Snapshot test with win 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-lg-8"
    >
      <div
        className="mt-4 card"
      >
        <div
          className="row"
        >
          <div
            className=" m-3 col-lg-12"
          >
            Next turn:   
            <span
              className="player"
            >
               Player 1: O 
            </span>
            <span
              className="player active"
            >
               Player 2: X
            </span>
          </div>
        </div>
      </div>
      <div
        className="mt-4 card"
      >
        <div
          className="wrapper"
        >
          <div
            className="board"
            id="board"
            onClick={[Function]}
          >
            <div
              className="tic-cell blt btt"
              id="0"
            >
              
            </div>
            <div
              className="tic-cell btt"
              id="1"
            >
              <svg
                className="circle piece"
              >
                <circle
                  cx="50%"
                  cy="50%"
                  fill="none"
                  r="45%"
                  stroke="#5a5a5a"
                  strokeWidth="3"
                />
              </svg>
              <svg
                className="strikeline"
              >
                <line
                  className="wl"
                  x1="48%"
                  x2="48%"
                  y1="0"
                  y2="100%"
                />
              </svg>
            </div>
            <div
              className="tic-cell btt brt"
              id="2"
            >
              
            </div>
            <div
              className="tic-cell blt"
              id="3"
            >
              <svg
                className="cross piece"
              >
                <line
                  stroke="#f47521"
                  strokeWidth="5"
                  x1="0%"
                  x2="100%"
                  y1="0"
                  y2="100%"
                />
                <line
                  stroke="#f47521"
                  strokeWidth="5"
                  x1="0"
                  x2="100%"
                  y1="100%"
                  y2="0"
                />
              </svg>
            </div>
            <div
              className="tic-cell"
              id="4"
            >
              <svg
                className="circle piece"
              >
                <circle
                  cx="50%"
                  cy="50%"
                  fill="none"
                  r="45%"
                  stroke="#5a5a5a"
                  strokeWidth="3"
                />
              </svg>
              <svg
                className="strikeline"
              >
                <line
                  className="wl"
                  x1="48%"
                  x2="48%"
                  y1="0"
                  y2="100%"
                />
              </svg>
            </div>
            <div
              className="tic-cell brt"
              id="5"
            >
              
            </div>
            <div
              className="tic-cell blt bbt"
              id="6"
            >
              <svg
                className="cross piece"
              >
                <line
                  stroke="#f47521"
                  strokeWidth="5"
                  x1="0%"
                  x2="100%"
                  y1="0"
                  y2="100%"
                />
                <line
                  stroke="#f47521"
                  strokeWidth="5"
                  x1="0"
                  x2="100%"
                  y1="100%"
                  y2="0"
                />
              </svg>
            </div>
            <div
              className="tic-cell bbt"
              id="7"
            >
              <svg
                className="circle piece"
              >
                <circle
                  cx="50%"
                  cy="50%"
                  fill="none"
                  r="45%"
                  stroke="#5a5a5a"
                  strokeWidth="3"
                />
              </svg>
              <svg
                className="strikeline"
              >
                <line
                  className="wl"
                  x1="48%"
                  x2="48%"
                  y1="0"
                  y2="100%"
                />
              </svg>
            </div>
            <div
              className="tic-cell bbt brt"
              id="8"
            >
              
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-md-4"
    >
      <div
        className="mt-4 card"
      >
        <div
          className="card-body"
        >
          <div
            className="card-title"
          >
            Game options:  
            <strong>
               
              Circles win!
            </strong>
          </div>
          <div
            className="input-group"
          >
            <div
              className="input-group-prepend"
            >
              <span
                className="input-group-text"
              >
                <input
                  aria-label="Checkbox for following text input"
                  className=""
                  onClick={[Function]}
                  type="checkbox"
                />
              </span>
            </div>
            <input
              className="form-control"
              disabled={true}
              placeholder="Zardoz Mode"
              type="text"
            />
          </div>
          <div
            className="pt-4 row"
          >
            <div
              className="col-4"
            >
              <button
                aria-label={null}
                className="btn btn-primary"
                onClick={[Function]}
                type="button"
              >
                 New game 
              </button>
            </div>
            <div
              className="col-4"
            >
              <button
                aria-label={null}
                className="btn btn-secondary"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                 Undo Last
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
